### PR TITLE
[API][SIMS #1134]Upgrade Typeorm - Part 1

### PR DESCRIPTION
**Pull request by guru-aot**
- [x]  Upgrade Typeorm to the latest version.
- [x] Upgraded dependent packages.
- [x] Refactoring find*, queries and connection, getConnection deprecated methods.
- [x] Restructured the type orm configuration.
- [x] Basic cleanup
- [x] Changed js configuration to ts

The version typeorm upgraded is 0.3.7 https://www.npmjs.com/package/typeorm/v/0.3.7 but the major changes on query and connection parameters happened on 0.3.2 https://www.npmjs.com/package/typeorm/v/0.3.2

### CODE REVIEW

File: sources/packages/api/env_setup.js
Comment (andrewsignori-aot) [inline sources/packages/api/env_setup.js:1]: As we talked lets revert these changes 😉

File: sources/packages/api/scripts/setupDB.ts
Comment (andrewsignori-aot) [inline sources/packages/api/scripts/setupDB.ts:1]: I believe that the pattern for the file naming would be `orm-config`.

File: sources/packages/api/scripts/setupDB.ts
Comment (andrewsignori-aot) [inline sources/packages/api/scripts/setupDB.ts:16]: Is it not better to create a Datasource for this migration file instead of reusing a global object?
Please consider that the process in this file runs completely decoupled from the rest of the API.

Comment (andrewsignori-aot) [inline sources/packages/api/scripts/setupDB.ts:16]: I believe that the destruction("destroy") must be moved to the `finally` of the `try/catch`.

File: sources/packages/api/src/app.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/app.service.ts:9]: Is there a reason to not have the `DataSource` injected here?

File: sources/packages/api/src/database/data.model.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/data.model.service.ts:7]: I believe that any can be replaced by `DataModel` here.

File: sources/packages/api/src/database/database.module.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/database.module.ts:9]: The schema is already set on `ormConfig`, can you please remove line 9 and directly use the `ormConfig` instead of the `finalConfig`?

File: sources/packages/api/src/database/database.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/database.service.ts:6]: Do we need this constructor?

File: sources/packages/api/src/services/application/application.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/application/application.service.ts:575]: Please add a TODO for further investigation on why `application.applicationStatus` was not translated to the DB column name. This will be taken care is the next PR, right?

File: sources/packages/api/src/services/application/application.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/application/application.service.ts:867]: Please add a TODO for further investigation on why `application.pirStatus` was not translated to the DB column name. This will be taken care is the next PR, right?

File: sources/packages/api/src/services/designation-agreement/designation-agreement-locations.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/designation-agreement/designation-agreement-locations.service.ts:14]: Please remove the `private readonly` as it is not necessary.

File: sources/packages/api/src/services/education-program/education-program.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/education-program/education-program.service.ts:244]: Please add a TODO for further investigation on why `programs.programStatus` was not translated to the DB column name. This will be taken care is the next PR, right?

File: sources/packages/api/src/services/education-program/education-program.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/education-program/education-program.service.ts:354]: Please add a TODO for further investigation on why `programs.programStatus` was not translated to the DB column name. This will be taken care is the next PR, right?

File: sources/packages/api/src/services/institution-type/institution-type.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution-type/institution-type.service.ts:11]: please remove the `private readonly` as it is not necessary here.

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:125]: Does the null value not affected by the below change? Can you please check it?
image
Source: https://github.com/typeorm/typeorm/blob/master/CHANGELOG.md

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:213]: Same comment from before. Should we use `IsNull()` instead?

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:509]: Same comment from before. Should we use `IsNull()` instead?

File: sources/packages/api/src/services/pir-denied-reason/pir-denied-reason.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/pir-denied-reason/pir-denied-reason.service.ts:8]: Please remove the `private readonly` as it is not necessary here.

File: sources/packages/api/src/services/program-year/program-year.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/program-year/program-year.service.ts:8]: Please remove the `private readonly` as it is not necessary here.

File: sources/packages/api/src/services/ssh/ssh.service.ts
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/ssh/ssh.service.ts:19]: Please rollback this change. The connection here means the SFTP connection, not the DB one.

File: sources/packages/api/src/services/application/application.service.ts
Comment (ann-aot) [inline sources/packages/api/src/services/application/application.service.ts:575]: why `application_status` instead of  `applicationStatus`?

File: sources/packages/api/tsconfig.build.json
Comment (andrewsignori-aot) [inline sources/packages/api/tsconfig.build.json:3]: Does `ormconfig.ts` need to be ignored? It is a typescript file to be compiled/transpiled as the others, right?

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/api/src/services/application/application.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/application/application.service.ts:575]: There is an unknown issue where apparently the model property is not translated to the column name. I asked @guru-aot to add a todo and do the investigation in the next PR right after this one. Do you see any concern about it @ann-aot?

Comment (ann-aot) [reply sources/packages/api/src/services/application/application.service.ts:575]: okay...no, I just want to know the reason.

**Review by ann-aot — COMMENTED** [review]
added a comment and some comments already asked by signori

File: sources/packages/api/scripts/setupDB.ts
Comment (dheepak-aot) [inline sources/packages/api/scripts/setupDB.ts:1]: Typeorm says as per their OOTB CLI setup, to declare the data source in data-source.ts.
image

File: sources/packages/api/src/app.controller.ts
Comment (dheepak-aot) [reply sources/packages/api/src/app.controller.ts:17]: Just to understand, what was the intention of this method when it was created?

File: sources/packages/api/scripts/setupDB.ts
Comment (dheepak-aot) [inline sources/packages/api/scripts/setupDB.ts:10]: simsDataSource.initialize() return the datasource instance not the connection. 
Hence the variable can be renamed to datasource instead of connection.

image

File: sources/packages/api/scripts/setupDB.ts
Comment (guru-aot) [reply sources/packages/api/scripts/setupDB.ts:1]: renaming to data-source.ts

File: sources/packages/api/src/database/config/ormconfig.ts
Comment (dheepak-aot) [inline sources/packages/api/src/database/config/ormconfig.ts:23]: Why did we remove this config? how will the typeorm know which directory to scan?

File: sources/packages/api/src/database/data.model.service.ts
Comment (guru-aot) [reply sources/packages/api/src/database/data.model.service.ts:7]: yes previous object definition was deprecated and used any as the closest one.

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (guru-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:125]: It did not throw any error or warnings, but thought of taking these changes in the next PR. But will change it as its comments are given now.

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (guru-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:509]: It did not throw any error or warnings, but thought of taking these changes in the next PR. But will change it as its comments are given now.

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (guru-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:213]: It did not throw any error or warnings, but thought of taking these changes in the next PR. But will change it as its comments are given now.

File: sources/packages/api/src/utilities/sqlLoader.ts
Comment (dheepak-aot) [inline sources/packages/api/src/utilities/sqlLoader.ts:10]: Sure that it works without this in windows also? I remember facing an issue while trying to remove this. I may be wrong but vaguely remember.

File: sources/packages/api/tsconfig.build.json
Comment (dheepak-aot) [reply sources/packages/api/tsconfig.build.json]: Valid point. Speaking beyond the context of the PR, IMO **/*spec.ts also shouldn't be excluded.

**Comment by dheepak-aot** [comment]
Excellent work @guru-aot 👍. Added some comments.

File: sources/packages/api/src/app.controller.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/app.controller.ts]: Probably return token info. Looking into it I do not see a reason to keep it.

File: sources/packages/api/src/services/institution/institution.service.ts
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/institution/institution.service.ts]: Asked for a change because it does not seems to be a recommendation but a "do it".
Thanks for doing the changes.

File: sources/packages/api/tsconfig.build.json
Comment (andrewsignori-aot) [reply sources/packages/api/tsconfig.build.json]: The spec files are not part of the PROD package, why should we keep them?
Sorry, I am not following.

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes, it is looking pretty good 👍 
I will just wait for the latest changes to give final approval.

**Review by henryoforeh-aot — APPROVED** [review]

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes @guru-aot 👍. LGTM.

File: docs/howto/db-migration.md
Comment (andrewsignori-aot) [inline docs/howto/db-migration.md:18]: ❤️

Comment (ann-aot) [reply docs/howto/db-migration.md:18]: +1

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changs, looks good 👍

**Review by ann-aot — APPROVED** [review]
Great Work @guru-aot  👍

**Pull request merged**
